clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -258,12 +258,12 @@ public synchronized int sendFetches() {
             if (log.isDebugEnabled()) {
                 log.debug("Sending {} {} to broker {}", isolationLevel, data.toString(), fetchTarget);
             }
-            RequestFuture<ClientResponse> future = client.send(fetchTarget, request);
-            // We add the node to the set of nodes with pending fetch requests before adding the
-            // listener because the future may have been fulfilled on another thread (e.g. during a
+            // We add the node to the set of nodes with pending fetch requests before sending the
+            // request to the client because the future may have been fulfilled on another thread (e.g. during a
             // disconnection being handled by the heartbeat thread) which will mean the listener
-            // will be invoked synchronously.
+            // will be invoked synchronously, and hence the added id would not be removed anymore.
             this.nodesWithPendingFetchRequests.add(entry.getKey().id());
+            RequestFuture<ClientResponse> future = client.send(fetchTarget, request);
```

Comment:
This is the actual fix, others are just log4j improvements I got during the debugging.